crates/bevy_ecs/src/storage/blob_array.rs

```diff
@@ -0,0 +1,455 @@
+use super::blob_vec::array_layout;
+use crate::storage::blob_vec::array_layout_unchecked;
+use bevy_ptr::{OwningPtr, Ptr, PtrMut};
+use bevy_utils::OnDrop;
+use std::{
+    alloc::{handle_alloc_error, Layout},
+    cell::UnsafeCell,
+    num::NonZeroUsize,
+    ptr::NonNull,
+};
+
+/// A flat, type-erased data storage type similar to a [`BlobVec`](super::blob_vec::BlobVec), but with the length and capacity cut out
+/// for performance reasons. This type is reliant on its owning type to store the capacity and length information.
+///
+/// Used to densely store homogeneous ECS data. A blob is usually just an arbitrary block of contiguous memory without any identity, and
+/// could be used to represent any arbitrary data (i.e. string, arrays, etc). This type only stores meta-data about the Blob that it stores,
+/// and a pointer to the location of the start of the array, similar to a C array.
+pub(super) struct BlobArray {
+    item_layout: Layout,
+    // the `data` ptr's layout is always `array_layout(item_layout, capacity)`
+    data: NonNull<u8>,
+    // None if the underlying type doesn't need to be dropped
+    pub drop: Option<unsafe fn(OwningPtr<'_>)>,
+}
```

Comment:
Is there a reason `ThinArrayPtr` has a `capacity` field for debug asserts while `BlobArray` does not?